R/discrete_normal.R

```diff
@@ -0,0 +1,142 @@
+      
+      # add breaks, vector of lower and upper bounds, and the lower and upper
+      # bound of supported values
+      # self$breaks <- breaks
+      # self$lower_bounds <- breaks[-length(breaks)]
+      # self$upper_bounds <- breaks[-1]
+      # self$lower_bound <- min(breaks)
+      # self$upper_bound <- max(breaks)
+      
+      # EXPERIMENTAL:
+      # convert breaks to edges, which
+      # will be used to gather samples at the breaks and convert them to 
+      # rounded values 
+      # while avoiding the use of round or floor, which assumes that 
+      # the breaks are all the integers
+      # perhaps more crucially, we use edges, not breaks, to integrate the CDFs
+      # because we say that the likelihood of observing a break value is the sum
+      # of probability from the lower to upper edges surrounding a break
+      # first, create the lower and upper bounds that will bin a continuous
+      # variable into breaks
+      # using midpoint between breaks for now (from diff)
+      # generate edges using midpoint, -Inf, and Inf
+      # breaks_diff <- diff(breaks)
+      # edges <- c(-Inf, breaks[-length(breaks)] + breaks_diff/2, Inf)
+      self$edges <- edges
+      self$lower_bounds <- edges[-length(edges)]
+      self$upper_bounds <- edges[-1]
+      self$lower_bound <- min(edges)
+      self$upper_bound <- max(edges)
+      
+      # add the nodes as parents and parameters
+      dim <- check_dims(mean, sd, target_dim = dim)
+      super$initialize("discrete_normal", dim, discrete = TRUE)
+      self$add_parameter(mean, "mean")
+      self$add_parameter(sd, "sd")
+      
+    },
+    
+    tf_distrib = function(parameters, dag) {
+      
+      mean <- parameters$mean
+      sd <- parameters$sd
+      
+      tf_breaks <- fl(self$breaks)
+      tf_edges <- fl(self$edges)
+      tf_lower_bounds <- fl(self$lower_bounds)
+      tf_upper_bounds <- fl(self$upper_bounds)
+      tf_lower_bound <- fl(self$lower_bound)
+      tf_upper_bound <- fl(self$upper_bound)
+      
+      log_prob <- function(x) {
+        
+        # build distribution object
+        d <- tfp$distributions$Normal(
+          loc = mean,
+          scale = sd
+        )
+        
+        # for those lumped into groups,
+        # compute the bounds of the observed groups
+        # and get tensors for the bounds in the format expected by TFP
+        tf_idx <- tfp$stats$find_bins(x, tf_edges)
+        tf_idx_int <- tf_as_integer(tf_idx)
+        tf_lower_vec <- tf$gather(tf_lower_bounds, tf_idx_int)
+        tf_upper_vec <- tf$gather(tf_upper_bounds, tf_idx_int)
+        
+        # compute the density over the observed groups
+        # note-to-self: this looks like https://mc-stan.org/docs/2_29/stan-users-guide/bayesian-measurement-error-model.html#rounding
+        low <- tf_safe_cdf(tf_lower_vec, d, tf_lower_bound, tf_upper_bound)
+        up <- tf_safe_cdf(tf_upper_vec, d, tf_lower_bound, tf_upper_bound)
+        log_density <- log(up - low)
```

Comment:
When the simulated data has high variance, e.g., `rnorm(..., sd = 10)`, I often get this error:
```
> draws <- mcmc(mod)
Error: Could not find reasonable starting values after 20 attempts.
Please specify initial values manually via the `initial_values` argument
```
The problem may lie in `log_density <- log(up - low)`. Gut feeling is that high variance generates data that are more widely spaced, so when `breaks` are regularly spaced integers (see `tf_idx`), some bins have zero elements, then `tf_safe_cdf` has nothing to evaluate? Then we end up taking the log of zero... ???